File: backend/src/main/java/com/rodrigocoelhoo/lifemanager/nutrition/service/IngredientService.java
package com.rodrigocoelhoo.lifemanager.nutrition.service;

import com.rodrigocoelhoo.lifemanager.config.RedisCacheService;
import com.rodrigocoelhoo.lifemanager.exceptions.ResourceNotFound;
import com.rodrigocoelhoo.lifemanager.nutrition.dto.IngredientDTO;
import com.rodrigocoelhoo.lifemanager.nutrition.dto.IngredientDetailsDTO;
import com.rodrigocoelhoo.lifemanager.nutrition.model.IngredientModel;
import com.rodrigocoelhoo.lifemanager.nutrition.repository.IngredientRepository;
import com.rodrigocoelhoo.lifemanager.users.UserModel;
import com.rodrigocoelhoo.lifemanager.users.UserService;
import jakarta.transaction.Transactional;
import jakarta.validation.Valid;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

@Service
public class IngredientService {

    private final IngredientRepository ingredientRepository;
    private final UserService userService;
    private final RedisCacheService redisCacheService;

    private static final String CACHE_LIST = "ingredients";

    public IngredientService(
            IngredientRepository ingredientRepository,
            UserService userService,
            RedisCacheService redisCacheService
    ) {
        this.ingredientRepository = ingredientRepository;
        this.userService = userService;
        this.redisCacheService = redisCacheService;
    }

    @Cacheable(value = CACHE_LIST, keyGenerator = "userAwareKeyGenerator")
    public Page<IngredientDetailsDTO> getAllIngredients(
            Pageable pageable,
            String name
    ) {
        UserModel user = userService.getLoggedInUser();
        Page<IngredientModel> page;
        if(name == null || name.isBlank())
            page = ingredientRepository.findAllByUser(user, pageable);
        else
            page = ingredientRepository.findByUserAndNameContainingIgnoreCase(user, name, pageable);

        return page.map(IngredientDetailsDTO::fromEntity);
    }

    public IngredientModel getIngredient(Long id) {
        UserModel user = userService.getLoggedInUser();
        return ingredientRepository.findByUserAndId(user, id)
                .orElseThrow(() -> new ResourceNotFound("Ingredient with ID '" + id + "' does not belong to the current user"));
    }

    @Transactional
    public IngredientModel createIngredient(IngredientDTO data) {
        UserModel user = userService.getLoggedInUser();

        IngredientModel ingredient = IngredientModel.builder()
                .name(data.name())
                .user(user)
                .brands(new HashSet<>())
                .build();

        IngredientModel saved =  ingredientRepository.save(ingredient);

        redisCacheService.evictUserCache(CACHE_LIST);

        return saved;
    }

    @Transactional
    public IngredientModel updateIngredient(Long id, @Valid IngredientDTO data) {
        IngredientModel ingredient = getIngredient(id);
        ingredient.setName(data.name());
        IngredientModel saved =  ingredientRepository.save(ingredient);

        redisCacheService.evictUserCache(CACHE_LIST);
        redisCacheService.evictUserCache("recipes");

        return saved;
    }

    @Transactional
    public void deleteIngredient(Long id) {
        IngredientModel ingredient = getIngredient(id);
        ingredientRepository.delete(ingredient);

        redisCacheService.evictUserCache(CACHE_LIST);
        redisCacheService.evictUserCache("recipes");
        redisCacheService.evictUserCache("meals");
        redisCacheService.evictUserCache("nutritionDashboard");
    }

    public List<IngredientModel> getIngredients(List<Long> ids) {
        if (ids == null || ids.isEmpty()) {
            return List.of();
        }

        UserModel user = userService.getLoggedInUser();
        List<IngredientModel> ingredients = ingredientRepository.findAllByUserAndIdIn(user, ids);

        if(ingredients.size() != ids.size())
            throw new ResourceNotFound("Some ingredients do not belong to the current user");

        return ingredients;
    }
}

File: frontend/src/services/nutrition/ingredient/ingredient.service.ts
import { api } from '../../api';
import type { PageResponseDTO } from '../../api.dto';
import type { IngredientDetailsDTO, IngredientDTO } from './ingredient.dto';

const BASE_URL = '/ingredients';

export const ingredientService = {

	getIngredients: async (
		page: number,
		size: number,
		name?: string
	): Promise<PageResponseDTO<IngredientDetailsDTO>> => {
		try {
			const params = new URLSearchParams({
				page: page.toString(),
				size: size.toString(),
			});

			if (name && name.trim() !== "") {
				params.append("name", name);
			}

			const { data } = await api.get<PageResponseDTO<IngredientDetailsDTO>>(`${BASE_URL}?${params.toString()}`);
			return data;
		} catch (error) {
			console.error('Failed to fetch ingredients:', error);
			throw new Error('Unable to retrieve ingredients. Please try again.');
		}
	},

	createIngredient: async (payload: IngredientDTO): Promise<IngredientDetailsDTO> => {
		try {
			const { data } = await api.post(BASE_URL, payload);
			return data;
		} catch (error) {
			console.error('Failed to create ingredient:', error);
			throw new Error('Unable to create ingredient. Please try again.');
		}
	},

	updateIngredient: async (
		id: number,
		payload: IngredientDTO
	): Promise<IngredientDetailsDTO> => {
		try {
			const { data } = await api.put(`${BASE_URL}/${id}`, payload);
			return data;
		} catch (error) {
			console.error(`Failed to update ingredient with id ${id}:`, error);
			throw new Error(`Unable to update ingredient with id ${id}:. Please try again.`);
		}
	},

	deleteIngredient: async (id: number): Promise<void> => {
		try {
			const { data } = await api.delete(`${BASE_URL}/${id}`);
			return data;
		} catch (error) {
			console.error(`Failed to delete ingredient with id ${id}:`, error);
			throw new Error(`Unable to delete ingredient with id ${id}:. Please try again.`);
		}
	},

}